src/include/imgpp/bcimgpp.hpp

```diff
@@ -0,0 +1,221 @@
+  }
+
+  uint32_t BlockWidth() const {
+    return bc_desc_.block_width;
+  }
+
+  uint32_t BlockHeight() const {
+    return bc_desc_.block_height;
+  }
+
+  uint32_t BlockBytes() const {
+    return bc_desc_.block_bytes;
+  }
+
+  uint32_t BlockPitch() const {
+    return block_pitch_;
+  }
+
+  uint32_t SlicePitch() const {
+    return slice_pitch_;
+  }
+
+  uint32_t HorizontalBlockNum() const {
+    return horizontal_block_num_;
+  }
+
+  uint32_t VerticalBlockNum() const {
+    return vertical_block_num;
+  }
+
+  const uint32_t *Dimensions () const {
+    return dimensions_;
+  }
+
+  const uint8_t *GetData() const {
+    return data_;
+  }
+
+  uint8_t *GetData() {
+    return data_;
+  }
+
+  static const uint32_t CalcBlockPitch(
+    BCFormat format, uint32_t w, uint8_t align_bytes = 1) {
+    auto &bc_desc = GetBCDesc(format);
+    if (format == Unknown || w % bc_desc.block_width != 0) {
+      return 0;
+    }
+    uint32_t num_of_width = w / bc_desc.block_width;
+    if (align_bytes) {
+      return (((num_of_width * bc_desc.block_bytes) >> 3) + (align_bytes - 1)) / align_bytes * align_bytes;
+    } else {
+      return (num_of_width * bc_desc.block_bytes) >> 3;
+    }
+  }
+
+  static const uint32_t CalcHorizontalBlockNum(BCFormat format, uint32_t w) {
+    auto &bc_desc = GetBCDesc(format);
+    return w / bc_desc.block_width;
+  }
+
+  static const uint32_t CalcVerticalBlockNum(BCFormat format, uint32_t h) {
+    auto &bc_desc = GetBCDesc(format);
+    return h / bc_desc.block_height;
+  }
+
+
+private:
+  void Validation() {
+    if (bc_desc_.format == Unknown || width_ % bc_desc_.block_width != 0 ||
+      height_ % bc_desc_.block_height != 0) {
+      Reset();
+    }
+  }
+
+  void Reset() {
+    data_ = nullptr;
+    width_ = height_ = depth_ = horizontal_block_num_ = vertical_block_num = block_pitch_ = slice_pitch_ = 0;
+  }
+
+  uint8_t *data_; //!<Data pointer. NOT a smart pointer hence NOT responsible the buffer!
+  BCDesc bc_desc_;
+
+  union {
+    struct {
+      uint32_t width_;
+      uint32_t height_;
+      uint32_t depth_;
+      uint32_t horizontal_block_num_;
+      uint32_t vertical_block_num;
+    };
+    uint32_t dimensions_[5];
+  };
+
+  uint32_t block_pitch_;
+  uint32_t slice_pitch_;
+};
+
+inline bool CopyData(BCImgROI &dst, const BCImgROI &src) {
+  if (src.Format() != dst.Format() || src.Width() > dst.Width() || src.Height() > dst.Height()
+    || src.Depth() > dst.Depth()) {
+    return false;
+  }
+
+  uint32_t src_length = BCImgROI::CalcBlockPitch(src.Format(), src.Width());
+  for (uint32_t layer = 0; layer < src.Depth(); ++layer) {
+    for (uint32_t y = 0; y < src.VerticalBlockNum(); ++y) {
+      memcpy(dst.BlockAt(0, y, layer), src.BlockAt(0, y, layer), src_length);
+    }
+  }
+  return true;
+}
+
+class BCImg: public ImgBase<BCImgROI, BCImg> {
+public:
+  BCImg() {}
+  ~BCImg() {}
+
+  BCImg(BCFormat format, uint32_t w, uint32_t h, uint32_t depth, uint8_t align_bytes = 1) {
```

Comment:
the four parameter constructor overload may be confused with following constructor